actors/evm/src/interpreter/instructions/memory.rs

```diff
@@ -53,6 +53,49 @@ pub fn get_memory_region(
     }))
 }
 
+#[inline]
+pub fn mcopy(
+    state: &mut ExecutionState,
+    _: &System<impl Runtime>,
+    dest_index: U256,
+    src_index: U256,
+    size: U256,
+) -> Result<(), ActorError> {
+    // Copy memory from src_index to dest_index.
+    // Handles overlapping slices as if using an intermediate buffer, ensuring correct copying.
+    // Expands memory if src_index + size or dest_index + size exceeds current bounds.
+    // Returns an error if memory regions are invalid or cannot be allocated.
+
+    copy_within_memory(&mut state.memory, dest_index, src_index, size)
+}
+
+pub fn copy_within_memory(
+    memory: &mut Memory,
+    dest_index: U256,
+    src_index: U256,
+    size: U256,
+) -> Result<(), ActorError> {
+    // Expand memory to accommodate requested src_index + size
+    let _region = get_memory_region(memory, src_index, size)?.expect("empty region");
+
+    // Expand memory to match dest_index + size
+    let _destination_region = get_memory_region(memory, dest_index, size)?.expect("empty region");
+
+    if size > 0 {
```

Comment:
Any reason not to do this check first to avoid the memory referencing?

Reply:
great idea! This fixes your concern I think b283a30d